clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/StrongServerStoreProxy.java

```diff
@@ -0,0 +1,149 @@
+ */
+public class StrongServerStoreProxy implements ServerStoreProxy {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(StrongServerStoreProxy.class);
+
+  private final ServerStoreProxy delegate;
+  private final ConcurrentMap<Long, CountDownLatch> invalidationsInProgress = new ConcurrentHashMap<Long, CountDownLatch>();
+  private final List<InvalidationListener> invalidationListeners = new CopyOnWriteArrayList<InvalidationListener>();
+
+  public StrongServerStoreProxy(String cacheId, final EhcacheClientEntity entity) {
+    this.delegate = new NoInvalidationServerStoreProxy(cacheId, entity);
+    entity.addResponseListener(EhcacheEntityResponse.InvalidationDone.class, new EhcacheClientEntity.ResponseListener<EhcacheEntityResponse.InvalidationDone>() {
+      @Override
+      public void onResponse(EhcacheEntityResponse.InvalidationDone response) {
+        if (response.getCacheId().equals(getCacheId())) {
+          long key = response.getKey();
+          LOGGER.debug("CLIENT: on cache {}, server notified that clients invalidated key {}", getCacheId(), key);
+          CountDownLatch countDownLatch = invalidationsInProgress.remove(key);
+          countDownLatch.countDown();
+        } else {
+          LOGGER.debug("CLIENT: on cache {}, ignoring invalidation on unrelated cache : {}", getCacheId(), response.getCacheId());
+        }
+      }
+    });
+    entity.addResponseListener(EhcacheEntityResponse.ClientInvalidateHash.class, new EhcacheClientEntity.ResponseListener<EhcacheEntityResponse.ClientInvalidateHash>() {
+      @Override
+      public void onResponse(EhcacheEntityResponse.ClientInvalidateHash response) {
+        final String cacheId = response.getCacheId();
+        final long key = response.getKey();
+        final int invalidationId = response.getInvalidationId();
+
+        LOGGER.debug("CLIENT: doing work to invalidate hash {} from cache {} (ID {})", key, cacheId, invalidationId);
+        for (InvalidationListener listener : invalidationListeners) {
+          listener.onInvalidationRequest(key);
+        }
+
+        try {
+          LOGGER.debug("CLIENT: ack'ing invalidation of hash {} from cache {} (ID {})", key, cacheId, invalidationId);
+          entity.invoke(EhcacheEntityMessage.clientInvalidateHashAck(cacheId, key, invalidationId), true);
+        } catch (Exception e) {
+          //TODO: what should be done here?
+          LOGGER.error("error acking client invalidation of hash " + key + " on cache " + cacheId, e);
+        }
+      }
+    });
+  }
+
+  private <T> T performWaitingForInvalidationIfNeeded(long key, NullaryFunction<T> c) throws InterruptedException {
+    CountDownLatch latch = new CountDownLatch(1);
+    while (true) {
+      CountDownLatch countDownLatch = invalidationsInProgress.putIfAbsent(key, latch);
+      if (countDownLatch == null) {
+        break;
+      }
+      countDownLatch.await();
+    }
+
+    try {
+      return c.apply();
+    } finally {
+      latch.await();
```

Comment:
This wait in finally seems risky. What if the operation failed to reach the server for example?
Also, I would rather have `await` with timeout and logging, even if we effectively end up waiting forever.